mlflow/utils/conda.py

```diff
@@ -84,12 +85,43 @@ def _get_conda_executable_for_create_env():
     return conda_env_create_path
 
 
-def _list_conda_environments():
-    prc = process._exec_cmd([get_conda_bin_executable("conda"), "env", "list", "--json"])
+def _list_conda_environments(extra_env=None):
+    prc = process._exec_cmd(
+        [get_conda_bin_executable("conda"), "env", "list", "--json"], extra_env=extra_env
+    )
     return list(map(os.path.basename, json.loads(prc.stdout).get("envs", [])))
 
 
-def get_or_create_conda_env(conda_env_path, env_id=None, capture_output=False):
+def _get_conda_extra_envs(env_root_dir=None):
+    if env_root_dir is not None:
+        # Create isolated conda package cache dir "conda_pkgs" under the env_root_dir
+        # for each python process.
+        # Note: shared conda package cache dir causes race condition issues:
+        # See https://github.com/conda/conda/issues/8870
+        # See https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/use-condarc.html#specify-environment-directories-envs-dirs
+        # and https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/use-condarc.html#specify-package-directories-pkgs-dirs
+
+        conda_env_path = os.path.join(env_root_dir, "conda_envs")
+        conda_pkgs_path = os.path.join(env_root_dir, "conda_pkgs")
+        pip_cache_dir = os.path.join(env_root_dir, "pip_pkgs")
+
+        os.makedirs(conda_env_path, exist_ok=True)
+        os.makedirs(conda_pkgs_path, exist_ok=True)
+        os.makedirs(pip_cache_dir, exist_ok=True)
+
+        return {
+            "CONDA_ENVS_PATH": conda_env_path,
+            "CONDA_PKGS_DIRS": conda_pkgs_path,
+            "PIP_CACHE_DIR": pip_cache_dir,
```

Comment:
Is this environment variable documented somewhere in pip's doc?

Reply:
See https://pip.pypa.io/en/stable/topics/configuration/#environment-variables

> pip’s command line options can be set with environment variables using the format PIP_<UPPER_LONG_NAME> . Dashes (-) have to be replaced with underscores (_).

Because pip has a command argument of "--cache-dir" , so setting env "PIP_CACHE_DIR" is equivalent to that argument according to the doc.

Reply:
got it!